<a href="https://colab.research.google.com/github/hend-22/-IT326-Project/blob/main/Phase2.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [56]:
!git clone https://github.com/hend-22/-IT326-Project.git

fatal: destination path '-IT326-Project' already exists and is not an empty directory.


In [57]:
!ls -la ./-IT326-Project

total 92
drwxr-xr-x 5 root root  4096 Sep 25 13:16 .
drwxr-xr-x 1 root root  4096 Sep 25 13:16 ..
drwxr-xr-x 2 root root  4096 Sep 25 13:16 Dataset
drwxr-xr-x 8 root root  4096 Sep 25 13:16 .git
-rw-r--r-- 1 root root 65679 Sep 25 13:16 Phase2.ipynb
-rw-r--r-- 1 root root  1788 Sep 25 13:16 README.md
drwxr-xr-x 2 root root  4096 Sep 25 13:16 Reports


In [58]:
!ls -la ./-IT326-Project/Dataset

total 4472
drwxr-xr-x 2 root root    4096 Sep 25 13:16 .
drwxr-xr-x 5 root root    4096 Sep 25 13:16 ..
-rw-r--r-- 1 root root 2134769 Sep 25 14:05 Preprocessed_dataset.csv
-rw-r--r-- 1 root root  819722 Sep 25 13:16 Raw_Dataset.csv
-rw-r--r-- 1 root root 1606101 Sep 25 13:16 Reasearch_Paper_Stress_Detection.pdf


In [59]:
import pandas as pd

file_path = "./-IT326-Project/Dataset/Raw_Dataset.csv"
df = pd.read_csv(file_path)

df.head()

,User_ID,Age,Gender,Physical_Activity_Level,Diet_Type,BMI,Caffeine_Intake_mg,Water_Intake_liters,Screen_Time_Hours,Daily_Steps,Calories_Burned,Resting_Heart_Rate,Daily_Sleep_Hours,Deep_Sleep_Hours,Sleep_Quality_Score,Daily_Stress_Level
0,U00001,56,Female,Low,Healthy,25.1,99.0,1.6,3.9,4731,1697,74.0,6.6,1.5,4.0,Low
1,U00002,69,Male,Low,Average,25.8,104.0,2.2,3.9,4440,1377,70.0,8.0,2.4,5.3,Low
2,U00003,46,Male,Low,Average,24.0,188.0,1.4,5.5,5023,1289,72.0,8.2,1.9,5.0,Medium
3,U00004,32,Female,Medium,Healthy,25.5,124.0,3.1,9.8,6082,1683,69.0,7.4,2.5,3.9,Low
4,U00005,60,Male,Medium,Unhealthy,NaN,0.0,2.4,5.3,7285,1621,73.0,7.9,2.2,5.2,Low


In [60]:
# Missing-value analysis
missing_count = df.isnull().sum()
missing_percentage = (df.isnull().sum() / len(df)) * 100

missing_analysis = pd.DataFrame({
    'Missing Values': missing_count,
    'Percentage (%)': missing_percentage
})

missing_analysis

,Missing Values,Percentage (%)
User_ID,0,0.00
Age,0,0.00
Gender,0,0.00
Physical_Activity_Level,0,0.00
Diet_Type,0,0.00
BMI,210,2.10
Caffeine_Intake_mg,206,2.06
Water_Intake_liters,0,0.00
Screen_Time_Hours,0,0.00
Daily_Steps,0,0.00


In [61]:
# Examine columns with missing values
missing_columns = df.columns[df.isnull().any()]

df[missing_columns].describe()

,BMI,Caffeine_Intake_mg,Resting_Heart_Rate,Daily_Sleep_Hours
count,9790.000000,9794.000000,9781.000000,9810.000000
mean,25.087406,207.365224,66.944177,6.957951
std,3.905276,137.681368,8.450527,0.907726
min,17.500000,0.000000,50.000000,3.500000
25%,22.300000,100.000000,61.000000,6.300000
50%,25.100000,201.000000,67.000000,7.000000
75%,27.800000,303.000000,73.000000,7.600000
max,38.000000,730.000000,95.000000,10.900000


In [62]:
# Check skewness of columns with missing values
df[missing_columns].skew()

,0
BMI,0.151747
Caffeine_Intake_mg,0.338752
Resting_Heart_Rate,-0.022946
Daily_Sleep_Hours,0.030105


In [63]:
# Preprocessing Technique #1: Mean Imputation

df_preprocessed = df.copy()

for column in missing_columns:
    df_preprocessed[column] = df_preprocessed[column].fillna(
        df_preprocessed[column].mean()
    )

# Check missing values after preprocessing
df_preprocessed[missing_columns].isnull().sum()

,0
BMI,0
Caffeine_Intake_mg,0
Resting_Heart_Rate,0
Daily_Sleep_Hours,0


In [64]:
# Compare missing values before and after preprocessing

comparison = pd.DataFrame({
    'Before': df[missing_columns].isnull().sum(),
    'After': df_preprocessed[missing_columns].isnull().sum()
})

comparison

,Before,After
BMI,210,0
Caffeine_Intake_mg,206,0
Resting_Heart_Rate,219,0
Daily_Sleep_Hours,190,0


In [65]:
# remove the user identifier
df_preprocessed = df_preprocessed.drop(columns=['User_ID'])

**User id is only an identifier and does not provide useful informatioon for analysis.**

In [66]:
# identify numerical attributes before converting categorical attributes
numeric_columns = df_preprocessed.select_dtypes(include='number').columns

In [74]:
# Preprocessing Technique #2: Normalization through Min-Max Scaling

from sklearn.preprocessing import MinMaxScaler
scaler = MinMaxScaler()
df_preprocessed[numeric_columns] = scaler.fit_transform(df_preprocessed[numeric_columns])

**numerical attributes have different ranges, so normalization was applied to put them on the same scale**

In [68]:
df_preprocessed.head()

,Age,Gender,Physical_Activity_Level,Diet_Type,BMI,Caffeine_Intake_mg,Water_Intake_liters,Screen_Time_Hours,Daily_Steps,Calories_Burned,Resting_Heart_Rate,Daily_Sleep_Hours,Deep_Sleep_Hours,Sleep_Quality_Score,Daily_Stress_Level
0,0.745098,Female,Low,Healthy,0.370732,0.135616,0.244444,0.158333,0.166184,0.500415,0.533333,0.418919,0.296296,0.461538,Low
1,1.000000,Male,Low,Average,0.404878,0.142466,0.377778,0.158333,0.153223,0.367635,0.444444,0.608108,0.629630,0.661538,Low
2,0.549020,Male,Low,Average,0.317073,0.257534,0.200000,0.291667,0.179190,0.331120,0.488889,0.635135,0.444444,0.615385,Medium
3,0.274510,Female,Medium,Healthy,0.390244,0.169863,0.577778,0.650000,0.226360,0.494606,0.422222,0.527027,0.666667,0.446154,Low
4,0.823529,Male,Medium,Unhealthy,0.370117,0.000000,0.422222,0.275000,0.279943,0.468880,0.511111,0.594595,0.555556,0.646154,Low


**All numerical attributes were scaled to a range of 0–1**

In [69]:
# Preprocessing Technique #3: Data Conversion through Label Encoding

from sklearn.preprocessing import LabelEncoder

le = LabelEncoder()

df_preprocessed['Gender'] = le.fit_transform(df_preprocessed['Gender'])
df_preprocessed['Physical_Activity_Level'] = le.fit_transform(df_preprocessed['Physical_Activity_Level'])
df_preprocessed['Diet_Type'] = le.fit_transform(df_preprocessed['Diet_Type'])
df_preprocessed['Daily_Stress_Level'] = le.fit_transform(df_preprocessed['Daily_Stress_Level'])

**Categorical attributes were represented as text and needed to be converted into numerical values for the data mining algorithms.**

In [70]:
df_preprocessed.head()

,Age,Gender,Physical_Activity_Level,Diet_Type,BMI,Caffeine_Intake_mg,Water_Intake_liters,Screen_Time_Hours,Daily_Steps,Calories_Burned,Resting_Heart_Rate,Daily_Sleep_Hours,Deep_Sleep_Hours,Sleep_Quality_Score,Daily_Stress_Level
0,0.745098,0,1,1,0.370732,0.135616,0.244444,0.158333,0.166184,0.500415,0.533333,0.418919,0.296296,0.461538,1
1,1.000000,1,1,0,0.404878,0.142466,0.377778,0.158333,0.153223,0.367635,0.444444,0.608108,0.629630,0.661538,1
2,0.549020,1,1,0,0.317073,0.257534,0.200000,0.291667,0.179190,0.331120,0.488889,0.635135,0.444444,0.615385,2
3,0.274510,0,2,1,0.390244,0.169863,0.577778,0.650000,0.226360,0.494606,0.422222,0.527027,0.666667,0.446154,1
4,0.823529,1,2,2,0.370117,0.000000,0.422222,0.275000,0.279943,0.468880,0.511111,0.594595,0.555556,0.646154,1


**Gender, Physical Activity Level, Diet Type, and Daily Stress Level were converted into numerical labels.**

In [77]:
# Comparison of raw and preprocessed datasets

comparison = pd.DataFrame(
    {
    '': ['Number of Records',
         'Number of Attributes',
         'Missing Values',
         'User_ID', 'Numerical Attributes', 'Categorical Attributes'],

    'Raw Dataset': [df.shape[0],
        df.shape[1],
        df.isnull().sum().sum(),
        'Present',
        'Original scale',
        'Text'],
    'Preprocessed Dataset':[
        df_preprocessed.shape[0],
        df_preprocessed.shape[1],
        df_preprocessed.isnull().sum().sum(),
        'Removed',
        'Normalized (0–1)',
        'Label encoded']
    })

display(comparison)

,,Raw Dataset,Preprocessed Dataset
0,Number of Records,10000,10000
1,Number of Attributes,16,15
2,Missing Values,825,0
3,User_ID,Present,Removed
4,Numerical Attributes,Original scale,Normalized (0–1)
5,Categorical Attributes,Text,Label encoded


In [78]:
print("Raw Dataset")
display(df.head())

print("Preprocessed Dataset")
display(df_preprocessed.head())

Raw Dataset


,User_ID,Age,Gender,Physical_Activity_Level,Diet_Type,BMI,Caffeine_Intake_mg,Water_Intake_liters,Screen_Time_Hours,Daily_Steps,Calories_Burned,Resting_Heart_Rate,Daily_Sleep_Hours,Deep_Sleep_Hours,Sleep_Quality_Score,Daily_Stress_Level
0,U00001,56,Female,Low,Healthy,25.1,99.0,1.6,3.9,4731,1697,74.0,6.6,1.5,4.0,Low
1,U00002,69,Male,Low,Average,25.8,104.0,2.2,3.9,4440,1377,70.0,8.0,2.4,5.3,Low
2,U00003,46,Male,Low,Average,24.0,188.0,1.4,5.5,5023,1289,72.0,8.2,1.9,5.0,Medium
3,U00004,32,Female,Medium,Healthy,25.5,124.0,3.1,9.8,6082,1683,69.0,7.4,2.5,3.9,Low
4,U00005,60,Male,Medium,Unhealthy,NaN,0.0,2.4,5.3,7285,1621,73.0,7.9,2.2,5.2,Low


Preprocessed Dataset


,Age,Gender,Physical_Activity_Level,Diet_Type,BMI,Caffeine_Intake_mg,Water_Intake_liters,Screen_Time_Hours,Daily_Steps,Calories_Burned,Resting_Heart_Rate,Daily_Sleep_Hours,Deep_Sleep_Hours,Sleep_Quality_Score,Daily_Stress_Level
0,0.745098,0,1,1,0.370732,0.135616,0.244444,0.158333,0.166184,0.500415,0.533333,0.418919,0.296296,0.461538,1
1,1.000000,1,1,0,0.404878,0.142466,0.377778,0.158333,0.153223,0.367635,0.444444,0.608108,0.629630,0.661538,1
2,0.549020,1,1,0,0.317073,0.257534,0.200000,0.291667,0.179190,0.331120,0.488889,0.635135,0.444444,0.615385,2
3,0.274510,0,2,1,0.390244,0.169863,0.577778,0.650000,0.226360,0.494606,0.422222,0.527027,0.666667,0.446154,1
4,0.823529,1,2,2,0.370117,0.000000,0.422222,0.275000,0.279943,0.468880,0.511111,0.594595,0.555556,0.646154,1


### Overall:
Missing values were handled using mean imputation, reducing the total number of missing values from 825 to 0. The numerical attributes were then normalized to a range of 0 to 1, while categorical attributes were converted into numerical values. The User_ID attribute was also removed because it is only an identifier. The resulting dataset contains 10,000 records and 15 attributes and was saved separately as Preprocessed_dataset.csv.

In [73]:
# Save the preprocessed dataset separately
output_path = "./-IT326-Project/Dataset/Preprocessed_dataset.csv"

df_preprocessed.to_csv(output_path, index=False)

print("Preprocessed dataset saved successfully.")

Preprocessed dataset saved successfully.


In [72]:
!ls -la ./-IT326-Project/Dataset

total 4472
drwxr-xr-x 2 root root    4096 Sep 25 13:16 .
drwxr-xr-x 5 root root    4096 Sep 25 13:16 ..
-rw-r--r-- 1 root root 2134769 Sep 25 14:10 Preprocessed_dataset.csv
-rw-r--r-- 1 root root  819722 Sep 25 13:16 Raw_Dataset.csv
-rw-r--r-- 1 root root 1606101 Sep 25 13:16 Reasearch_Paper_Stress_Detection.pdf


The dataset was analyzed for missing values by calculating the number and percentage of missing values in each attribute.
